Networking

VitaliiDiadchenko opened the pull request:
it added load data from a network

VitaliiDiadchenko committed:
Created BuildConfig
Implementation Retrofit and okhttp libraries
Created MovieApiFactory.kt, MovieApiService.kt and SerializeService.kt
Removed LoadData.kt

VitaliiDiadchenko committed:
Created MovieDetailViewModel.kt and MovieDetailViewModelFactory.kt
Added parameter List<Actor> in ActorAdapter
Renamed adapters
Added Interceptor

an-yurii commented on app/build.gradle:
```diff
@@ -53,12 +69,19 @@ dependencies {
+    implementation 'com.jakewharton.retrofit:retrofit2-kotlinx-serialization-converter:0.8.0'
+
+    //okhttp3 library
+    implementation 'com.squareup.okhttp3:okhttp:4.9.0'
```
Библиотеки okhttp3 также имею одну версию, как и lifecycle

an-yurii commented on app/build.gradle:
```diff
@@ -21,11 +21,27 @@ android {
         }
     }
+
+    flavorDimensions "serverType"
```
Для нашего проекта - излишне. Достаточно воспользоваться buildType
https://t.me/aa_fundamentals_0/696

an-yurii commented on app/src/main/java/com/vitaliidiadchenko/mykotkinapplication/networkService/ApiKeyInterceptor.kt:
```diff
@@ -0,0 +1,17 @@
+package com.vitaliidiadchenko.mykotkinapplication.networkService
```
В Андроид service - одна из четырех базовых сущностей. Лучше избегать этого слова в своих названиях, чтобы не путаться.

an-yurii commented on app/src/main/java/com/vitaliidiadchenko/mykotkinapplication/networkService/MovieApiFactory.kt:
```diff
@@ -0,0 +1,39 @@
+        .client(client)
+        .addConverterFactory(json.asConverterFactory("application/json".toMediaType()))
+        .build()
+
+
+
+
+
+
+
+
```
Очищай пустые строки. Держи код в порядке.

an-yurii commented on app/src/main/java/com/vitaliidiadchenko/mykotkinapplication/networkService/MovieApiFactory.kt:
```diff
@@ -0,0 +1,39 @@
+        .build()
+
+    @ExperimentalSerializationApi
+    val retrofit: Retrofit = Retrofit.Builder()
```
Это - холдер, не фабрика. Фабрика каждый раз создает объект.

У тебя получилась "жадная" инициализация переменных. Переменные будут инициализированы во время загрузки класса. Это увеличивает время запуска приложения. Переделай на ленивую инициализацию.
```
object RetrofitHolder {

    @ExperimentalSerializationApi
    val retrofit: Retrofit by lazy {
        Retrofit.Builder()
            .client(createClient())
            .baseUrl(BuildConfig.BASE_URL)
            .addConverterFactory(createJson().asConverterFactory("application/json".toMediaType()))
            .build()
    }

    private fun createClient() = OkHttpClient().newBuilder()
        .addInterceptor(HttpLoggingInterceptor().setLevel(HttpLoggingInterceptor.Level.BODY))
        .addInterceptor(MovieApiHeaderInterceptor())
        .build()

    private fun createJson() = Json {
        ignoreUnknownKeys = true
    }

}
```

an-yurii commented on app/src/main/java/com/vitaliidiadchenko/mykotkinapplication/networkService/MovieApiService.kt:
```diff
@@ -0,0 +1,18 @@
+interface MovieApiService {
+
+    @GET("movie/popular")
+    suspend fun getMovies(): MoviesPojo
```
Pojo - это пережиток Java. Котлин нам генерирует геттеры и сеттры, поэтому отпала надобность в pojo объектах. Во вторых, нужно стараться избегать использования названий, которые говорят об архитектуре. Нужно использовать те, которые говорят о назначении класса. В данном случае, речь идет о Dto (data transfer object), объектах сетевого слоя.

an-yurii commented on app/src/main/java/com/vitaliidiadchenko/mykotkinapplication/networkService/SerializeService.kt:
```diff
@@ -0,0 +1,88 @@
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class MoviesPojo(
```
`MoviesPojo` -> `MoviesDto`
Тоже самое и для остальных классов

an-yurii commented on app/src/main/java/com/vitaliidiadchenko/mykotkinapplication/networkService/SerializeService.kt:
```diff
@@ -0,0 +1,88 @@
+    val name: String
+)
+//method to deserialize ListMovies
+suspend fun deserializeMovies(
```
Избегай "спихивания" вхего подряд в один класс. Он превращается в помойку и усложняется навигация по коду без IDE. Этот файл нужно разбить на четыре: модели для загрузки фильмов и актеров, мапперы для списка фильмов и актеров.

an-yurii commented on app/src/main/java/com/vitaliidiadchenko/mykotkinapplication/networkService/SerializeService.kt:
```diff
@@ -0,0 +1,88 @@
+            adult = moviesPojo.adult,
+            runtime = moviesPojo.runtime,
+            genres = moviesPojo.genreIds.map {
+                genresMap[it]?.name.toString()
```
Вызов `.toString()` тут излишен. name -, и так, строка

an-yurii commented on app/src/main/java/com/vitaliidiadchenko/mykotkinapplication/screens/movieDetail/ActorAdapter.kt:
```diff
@@ -45,10 +42,11 @@ class ActorViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
             .circleCrop()
-        Glide.with(context).load(actor.picture).apply(imageOption).into(avatar)
+        Glide.with(context)
+            .load(BuildConfig.BASE_IMAGE_URL + BuildConfig.ACTOR_AVATAR_SIZE + actor.picture)
```
Преобразование параметров нужно перенести в маппер из ActorDto в Actor. В маппере эту логику будет легко протестировать, вью будет оставаться максимально простой.
https://t.me/aa_fundamentals_0/698

an-yurii commented on app/src/main/java/com/vitaliidiadchenko/mykotkinapplication/screens/movieDetail/ActorAdapter.kt:
```diff
@@ -45,10 +42,11 @@ class ActorViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
-        Glide.with(context).load(actor.picture).apply(imageOption).into(avatar)
+        Glide.with(context)
+            .load(BuildConfig.BASE_IMAGE_URL + BuildConfig.ACTOR_AVATAR_SIZE + actor.picture)
+            .apply(imageOption).into(avatar)
```
```suggestion
            .apply(imageOption)
            .into(avatar)
```

an-yurii commented on app/src/main/java/com/vitaliidiadchenko/mykotkinapplication/screens/movieDetail/MovieDetailFragment.kt:
```diff
@@ -64,12 +70,12 @@ class MovieDetailFragment : Fragment() {
             val backdrop = findViewById<ImageView>(R.id.image_poster)
             Glide.with(context)
-                .load(movie.backdrop)
+                .load(BuildConfig.BASE_IMAGE_URL + BuildConfig.BIG_POSTER_SIZE + movie.backdrop)
```
Вью вынуждена знать о способе получения URL. Нужно перенести формирование URL в маппер https://t.me/aa_fundamentals_0/698

an-yurii commented on app/src/main/java/com/vitaliidiadchenko/mykotkinapplication/screens/movieDetail/MovieDetailFragment.kt:
```diff
@@ -81,14 +87,18 @@ class MovieDetailFragment : Fragment() {
+    //set actor's images to recyclerView
+    private fun setupActors(actors: List<Actor>) {
+        if (actors.isNotEmpty()) {
+            view?.apply {
```
Правильнее использовать идиому https://kotlinlang.org/docs/idioms.html#execute-if-not-null
```
view?.let {
    it.findViewById<TextView>(R.id.text_cast).isVisible = true
    ...
}
```

VitaliiDiadchenko committed:
fix implementation okhttp library
added buildConfigField to defaultConfig
Rename package networkService to networkModule
Rename Pojo to Dto
fix fun apply to let at initialization view in DetailFragment
fix mistake in MovieDetailViewModel at initialization mutableLivedata
fix mistake in BASE_URL

VitaliiDiadchenko committed:
Rename MovieApiFactory.kt to MovieApiHolder.kt

VitaliiDiadchenko committed:
Fix mistakes in MovieDetailViewModelFactory.kt and MovieListViewModelFactory.kt

an-yurii commented on app/src/main/java/com/vitaliidiadchenko/mykotkinapplication/networkModule/Dto/DeserializeMovies.kt:
```diff
@@ -0,0 +1,30 @@
+import kotlinx.coroutines.withContext
+
+//method to deserialize ListMovies
+suspend fun deserializeMovies(
```
Это - функция маппер, не десериализатор. Десериализацию делает плагин к ретрофиту.

an-yurii commented on app/src/main/java/com/vitaliidiadchenko/mykotkinapplication/networkModule/Dto/DeseriolizeActors.kt:
```diff
@@ -0,0 +1,20 @@
+import kotlinx.coroutines.withContext
+
+//method to deserialize ListActors
+suspend fun deserializeActors(
```
Это - функция маппер, не десериализатор. Десериализацию делает плагин к ретрофиту.

an-yurii commented on app/src/main/java/com/vitaliidiadchenko/mykotkinapplication/networkModule/MovieApiHolder.kt:
```diff
@@ -0,0 +1,32 @@
+    private val loggingInterceptor = HttpLoggingInterceptor().apply {
+        level = HttpLoggingInterceptor.Level.BODY
+    }
+    private val client = OkHttpClient.Builder()
```
Жадная инициализация переменной. Или переделай на ленивую, или переделай в метод.

an-yurii commented on app/src/main/java/com/vitaliidiadchenko/mykotkinapplication/networkModule/MovieApiHolder.kt:
```diff
@@ -0,0 +1,32 @@
+
+object MovieApiHolder {
+
+    private val json = Json { ignoreUnknownKeys = true }
```
Жадная инициализация переменной. Или переделай на ленивую, или переделай в метод.

an-yurii commented on app/src/main/java/com/vitaliidiadchenko/mykotkinapplication/networkModule/MovieApiHolder.kt:
```diff
@@ -0,0 +1,32 @@
+
+    private val json = Json { ignoreUnknownKeys = true }
+
+    private val loggingInterceptor = HttpLoggingInterceptor().apply {
```
Можно было не выносить в переменную.

an-yurii commented on app/src/main/java/com/vitaliidiadchenko/mykotkinapplication/networkModule/MovieApiHolder.kt:
```diff
@@ -0,0 +1,32 @@
+import okhttp3.logging.HttpLoggingInterceptor
+import retrofit2.Retrofit
+
+object MovieApiHolder {
```
MovieApiHolder -> RetrofitHolder

an-yurii commented on app/src/main/java/com/vitaliidiadchenko/mykotkinapplication/networkModule/MovieApiService.kt:
```diff
@@ -0,0 +1,21 @@
+import retrofit2.http.GET
+import retrofit2.http.Path
+
+interface MovieApiService {
```
Принято называть просто MovieApi. Service - компонент Андроид.

VitaliiDiadchenko committed:
Rename functions deserializeMovies and deserializeActors
Rename MovieApiHolder.kt to RetrofitHolder.kt
Create private val client by lazy in RetrofitHolder.kt
Added initialization Json { ignoreUnknownKeys = true } in RetrofitBuilder
Rename MovieApiService.kt to MovieApi.kt
Added initialization loggingInterceptor in OkHttpClient Builder
